### DISCUSSION TITLE
Should I pass `py::list` or `py::list&` to my function

### DISCUSSION THREAD

**Discussion by kriscao-cohere** [pybind/pybind11#5379] [Q&A] [answered]
I'm trying to do some processing on a long Python list of long strings in C++, and so I have a C++ object whose constructor I'd like to take in a Python list, get an owning reference to it, and then start asynchronously doing processing on the list with the GIL released. However, I don't know whether I should type the Python list in the C++ constructor as `py::list` or `py::list&`.

Theoretically, a `py::list` should just be a pointer to a Python object in heap memory, so both should be memory equivalent, is my understanding correct? However, I don't quite understand how reference counting works, so I don't know which one would increment the ref count correctly for the lifetime of my C++ object.

**Comment by Ishu-opps** [comment, marked as answer]
Try py::list
